tests

rjarry opened the pull request:
- **build: add makefile**
- **src: remove redundant ecoli_ prefix**
- **strvec: add missing ecoli_utils.h include**
- **file: fix completion errors when file does not exist**
- **mk: add tests and coverage targets**
- **examples: move test-completion.sh there**
- **node_file: allow changing functions**
- **tests: split into multiple files**
- **malloc: remove redundant libc wrappers**
- **github: remove codeql workflow**
- **github: enable test coverage in CI**

rjarry committed:
build: add makefile

Add a makefile for developer quality of life. By default `make` will run
meson setup and ninja with sensible arguments.

Include targets for running tests and generating coverage reports.

The 'tests' target runs meson test with error logs printed. The
'coverage' target depends on tests and uses gcovr to generate both HTML
and text coverage reports, excluding test/ and examples/ directories.
The coverage option is passed to meson via -Db_coverage flag.

The coverage report is output to $(BUILDDIR)/coverage/index.html.

Signed-off-by: Robin Jarry <rjarry@redhat.com>
Acked-by: Olivier Matz <zer0@droids-corp.org>

rjarry committed:
strvec: add missing ecoli_utils.h include

This header is needed for the EC_VA_END macro used in the header file.

Signed-off-by: Robin Jarry <rjarry@redhat.com>
Acked-by: Olivier Matz <zer0@droids-corp.org>

rjarry committed:
file: fix completion errors when file does not exist

There is no point in failing the ec_complete() call completely when
trying to complete a non-existent file.

Fix all error conditions accordingly.

Signed-off-by: Robin Jarry <rjarry@redhat.com>
Acked-by: Olivier Matz <zer0@droids-corp.org>

rjarry committed:
malloc: remove redundant libc wrappers

Remove the custom malloc wrapper layer and use standard libc allocation
functions directly. These wrappers were added before knowing about
libasan and are now redundant since we can compile with
-Db_sanitize=address to achieve similar or better functionality.

Delete ecoli_malloc.h and remove ec_malloc_register() infrastructure.
Replace all ec_malloc(), ec_free(), ec_realloc(), ec_calloc(),
ec_strdup(), and ec_strndup() calls with their standard libc equivalents
throughout the codebase.

Signed-off-by: Robin Jarry <rjarry@redhat.com>
Acked-by: Olivier Matz <zer0@droids-corp.org>

rjarry committed:
examples: move test-completion.sh there

Move test-completion.sh from the project root to the examples directory.
This script demonstrates how to use the completion functionality and
belongs with other example code rather than in the top level of the
repository.

Signed-off-by: Robin Jarry <rjarry@redhat.com>
Acked-by: Olivier Matz <zer0@droids-corp.org>

rjarry committed:
node_file: allow changing functions

Refactor ec_node_file to use a global operations table instead of
per-node function pointers. This allows tests to inject custom
filesystem operations via ec_node_file_set_ops() without accessing
private node data.

Move ec_node_file_ops to the public header and eliminate the private
structure and init_priv callback.

This will be used in the next commit when we split test code in separate
files.

Signed-off-by: Robin Jarry <rjarry@redhat.com>
Acked-by: Olivier Matz <zer0@droids-corp.org>

rjarry committed:
tests: split into multiple files

Extract embedded test code from library source files into a separate
test directory. Remove the EC_TEST_REGISTER infrastructure and all
LCOV_EXCL markers around test functions. This separates production code
from test code, improving maintainability.

Remove ecoli_test.h from the public API. Export internal symbols needed
by tests (ec_log_fct, ec_log_opaque, ec_log_lookup). Move coverage
compilation flags from meson to the Makefile where they are controlled
by the COVERAGE variable.

Signed-off-by: Robin Jarry <rjarry@redhat.com>
Acked-by: Olivier Matz <zer0@droids-corp.org>

rjarry committed:
github: remove codeql workflow

This is not used anymore.

Signed-off-by: Robin Jarry <rjarry@redhat.com>
Acked-by: Olivier Matz <zer0@droids-corp.org>

rjarry committed:
github: enable test coverage in CI

Install gcovr and build using the makefile.

Signed-off-by: Robin Jarry <rjarry@redhat.com>
Acked-by: Olivier Matz <zer0@droids-corp.org>

zer0-droids reviewed (commented):
[src: remove redundant ecoli_ prefix](https://github.com/rjarry/libecoli/pull/11/commits/ce5581d4e004b08f2538af96473d0f794c27aa8d)

The initial idea for file naming was :
- for header files, we use a prefixed name to avoid namespace conflict
- for C files, use the same name than header file
I'm not sure this patch really makes the situation clearer. What is the problem of having a prefix for C files?

zer0-droids reviewed (commented):
[build: add makefile](https://github.com/rjarry/libecoli/pull/11/commits/f4f48598deb4cda234a97e90821d3e3217834f41)

Would it make sense to add a simple documentation? Either in a "make help" target or in an README

zer0-droids reviewed (commented):
[examples: move test-completion.sh there](https://github.com/rjarry/libecoli/pull/11/commits/d2f4cbda82c918556ddf2b66d17fb6960317abdb)

I'll tend to move the example in examples/parse-yaml

zer0-droids reviewed (commented):
[node_file: allow changing functions](https://github.com/rjarry/libecoli/pull/11/commits/0350cdd360f9728eaef860b7a565205c65d3e9f3)

I don't understand, wasn't it possible to simply export ec_node_file_override_functions() ?

zer0-droids reviewed (commented):
[malloc: remove redundant libc wrappers](https://github.com/rjarry/libecoli/pull/11/commits/ae8bbbae2b2c8f3a641295e443fa255ca0d3cf05)

I'm quite undecided for this one. Having an allocation wrapper helped me to debug a leak in production code, because it is easier to segregate allocations of the lib from other applications. This is quite common to have this kind of wrapper in a library. I could help someone wanting to run the library in a non standard environment (baremetal without OS ?).
On the other hand, it removes a wrapper and the memory tracking feature used in tests is replaceable by asan.

rjarry commented:
> [build: add makefile](https://github.com/rjarry/libecoli/pull/11/commits/f4f48598deb4cda234a97e90821d3e3217834f41)
> 
> Would it make sense to add a simple documentation? Either in a "make help" target or in an README

It seemed self explanatory to me since you have bash completion with make targets. I can add some comments above the make targets for reference, if you want?

rjarry commented:
> [node_file: allow changing functions](https://github.com/rjarry/libecoli/pull/11/commits/0350cdd360f9728eaef860b7a565205c65d3e9f3)
> 
> I don't understand, wasn't it possible to simply export ec_node_file_override_functions() ?

The issue is that ec_node_file_override_functions() takes no parameter. And since I want to split the sources from the test code, I needed to pass the function pointers somehow. I didn't want to expose the private node info since it didn't make much sense. It didn't make much sense to have per-node special function pointers neither, so I replaced that with a global static variable/structure only used in tests.

rjarry commented:
> [src: remove redundant ecoli_ prefix](https://github.com/rjarry/libecoli/pull/11/commits/ce5581d4e004b08f2538af96473d0f794c27aa8d)
> 
> The initial idea for file naming was :
> 
> * for header files, we use a prefixed name to avoid namespace conflict
> * for C files, use the same name than header file
>   I'm not sure this patch really makes the situation clearer. What is the problem of having a prefix for C files?

I found it rather ugly for C files since they are not visible from the outside. I usually don't like when *all files* from a single folder have the same prefix. It usually tells "you should put them all into a subfolder" to me.

Do you really find this confusing?

rjarry commented:
> [examples: move test-completion.sh there](https://github.com/rjarry/libecoli/pull/11/commits/d2f4cbda82c918556ddf2b66d17fb6960317abdb)
> 
> I'll tend to move the example in examples/parse-yaml

Ack

rjarry commented:
> [malloc: remove redundant libc wrappers](https://github.com/rjarry/libecoli/pull/11/commits/ae8bbbae2b2c8f3a641295e443fa255ca0d3cf05)
> 
> I'm quite undecided for this one. Having an allocation wrapper helped me to debug a leak in production code, because it is easier to segregate allocations of the lib from other applications. This is quite common to have this kind of wrapper in a library. I could help someone wanting to run the library in a non standard environment (baremetal without OS ?). On the other hand, it removes a wrapper and the memory tracking feature used in tests is replaceable by asan.

My reasoning was that there are a few places with direct calls to `malloc` and `free` (e.g. `ec_node_file` accesses stuff from the libc which needs to be managed by the raw functions, same thing with libedit stuff). This introduced a confusion: which functions should be called and when? Replacing everything with plain malloc/free removes any possible confusion.

Also, do you really think this lib could be used on baremetal?

zer0-droids commented:
Hello @rjarry,

The file src/ecoli_node_expr_test.c remains (with diff conficts artefacts) after this commit : [tests: split into multiple files](https://github.com/rjarry/libecoli/pull/11/commits/fe53b0b44035bb4f9c520f8539b4471119984fb1)
[fe53b0b](https://github.com/rjarry/libecoli/pull/11/commits/fe53b0b44035bb4f9c520f8539b4471119984fb1)

rjarry commented:
> The file src/ecoli_node_expr_test.c remains (with diff conficts artefacts) after this commit : [tests: split into multiple files](https://github.com/rjarry/libecoli/pull/11/commits/fe53b0b44035bb4f9c520f8539b4471119984fb1) [fe53b0b](https://github.com/rjarry/libecoli/pull/11/commits/fe53b0b44035bb4f9c520f8539b4471119984fb1)

Nice catch! I fixed it, thanks.

zer0-droids commented:
mk: add tests and coverage targets
==================================

```
--- a/Makefile
+++ b/Makefile
@@ -42,7 +61,7 @@ tag-release:
 	git commit -sm "release: version $$next_version" -m "`git shortlog -n v$$cur_version..`" meson.build && \
 	git tag -sm "`git shortlog -n v$$cur_version..HEAD^`" "v$$next_version"
 
-.PHONY: all
+.PHONY: help
 help:
 	$Q echo 'Available targets:'
 	$Q echo
```

is it expected in this patch? I guess it should be in the previous one.

examples: move test-completion.sh there
=======================================

```
From 24f5993b6762919a987df3c77a057c2a67338e33 Mon Sep 17 00:00:00 2001
From: Robin Jarry <rjarry@redhat.com>
Date: Wed, 15 Oct 2025 13:31:03 +0200
Subject: [PATCH] examples: move test-completion.sh there

Move test-completion.sh from the project root to the examples directory.
This script demonstrates how to use the completion functionality and
belongs with other example code rather than in the top level of the
repository.

Signed-off-by: Robin Jarry <rjarry@redhat.com>
---
 test-completion.sh => examples/parse-yaml/test-completion.sh | 0
 1 file changed, 0 insertions(+), 0 deletions(-)
 rename test-completion.sh => examples/parse-yaml/test-completion.sh (100%)

diff --git a/test-completion.sh b/examples/parse-yaml/test-completion.sh
similarity index 100%
rename from test-completion.sh
rename to examples/parse-yaml/test-completion.sh
```

Just a side node: it seems anyway that the test-completion.sh script is broken and incomplete. I'll send a PR to fix this.

node_file: allow changing functions
===================================

```
--- a/src/ecoli_node_file.c
+++ b/src/ecoli_node_file.c
@@ -2,6 +2,7 @@
  * Copyright 2016, Olivier MATZ <zer0@droids-corp.org>
  */
 
+#include <assert.h>
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <stdio.h>
@@ -23,17 +24,20 @@
 
 EC_LOG_TYPE_REGISTER(node_file);
 
-struct ec_node_file {
-	/* below functions pointers are only useful for test */
-	int (*lstat)(const char *pathname, struct stat *buf);
-	DIR *(*opendir)(const char *name);
-	struct dirent *(*readdir)(DIR *dirp);
-	int (*closedir)(DIR *dirp);
-	int (*dirfd)(DIR *dirp);
-	int (*fstatat)(int dirfd, const char *pathname, struct stat *buf,
-		int flags);
+static struct ec_node_file_ops file_ops = {
+	.lstat = lstat,
+	.opendir = opendir,
+	.readdir = readdir,
+	.closedir = closedir,
+	.dirfd = dirfd,
+	.fstatat = fstatat,
 };
 
+void ec_node_file_set_ops(const struct ec_node_file_ops *ops) {
+	assert(ops != NULL);
+	file_ops = *ops;
+}
+
```

detail, but the opening bracket is not on a new line. Looking forward to the clang-format PR ;)

After fixing these 2 minor things, you can add my ack. Thanks @rjarry!

rjarry commented:
Done! I have squashed the first two commits. After some thought, they didn't make much sense split appart. I have moved the opening bracket to its own line as god meant it to be and fixed an unused result warning about asprintf.
